181/replan architecture to be testable

EmilKronholm opened the pull request:
# Big replan and restructure

This PR re structures and refactors the entire server. Files are moved around, and folders are moved around. No real new features are added.

## Important things and known shortcomings.
- To authenticate, you pass token as a parameter. This is bad practice and will be changed in future.
- If no token is passed, you will authenticate automatically.
- Streaming is not being tested.
- You need an .env file for this right now (with non-secret values). See discord `#backend`.
- There is documentation for the server in the `docs` folder.

We are not 100% satisfied for the server as we thing it is a bit headache to work with. If you have any ideas on how to restructure it again and improve it, we would love to hear it.

## New tests
We have now written test for the server. To run it write `npm test`.


## Start the server
To start the server run `node server.js`

athinakantis reviewed (approved):
### **Tested routes:** 

`/`: Works fine
`/login`: It did generate me a token when passing the correct data. Gives appropriate error message when entering incorrect data.
`/images`: Works without a token. However even when providing token generated during login, it returned 401 and the reason was that `token.userId is not defined`. Maybe this is on me.
`/images/:name`: Returned a (broken) image without a token.

### **Tested unit tests:**
I ran the tests created in /tests folder, and all passed, however gave me this error:
```
Jest did not exit one second after the test run has completed.

'This usually means that there are asynchronous operations that weren't stopped in your tests. Consider running Jest with `--detectOpenHandles` to troubleshoot this issue.
```
Keep in mind I commented out the `startStreamingServer()` function in `app.js`, if that's the reason why, then ignore this.

### **Summary**
This review does not really cover the streaming or messaging functions of this PR. However, I think the authentication looks good enough for now. Some things to consider:

1. Make sure to add .env to gitignore
2. Maybe consider removing some console.logs
3. Consider which routes are going to need authentication and which are not

Keep in mind this is not fully my area of expertise, but hopefully something here was helpful!

EmilKronholm commented on Backend/Server/api/routers/imageRouter.js:
```diff
@@ -1,46 +1,49 @@
+});
 
-expressApp.get('/images', (req, res) => {
+router.get('/images', authCheck, (req, res) => {
```
Good point, would definitly make it more structured. I think all routes need authentication but I will double check it, if so then we could do it!

EmilKronholm commented on Backend/Server/api/routers/userRouter.js:
Ja, det är en bra idé! Jag vet dock inte hur det fungerar i mobilappen, vad jag vet så, till skillnad från webbläsare, finns ingen automatisk cookies system så tror man manuellt måste skicka med ett field varje request med token. 

Jag ska kika vad som är enklast på android :) Tack för tipset!

EmilKronholm commented:
> ### **Tested routes:**
> `/`: Works fine `/login`: It did generate me a token when passing the correct data. Gives appropriate error message when entering incorrect data. `/images`: Works without a token. However even when providing token generated during login, it returned 401 and the reason was that `token.userId is not defined`. Maybe this is on me. `/images/:name`: Returned a (broken) image without a token.
> 
> ### **Tested unit tests:**
> I ran the tests created in /tests folder, and all passed, however gave me this error:
> 
> ```
> Jest did not exit one second after the test run has completed.
> 
> 'This usually means that there are asynchronous operations that weren't stopped in your tests. Consider running Jest with `--detectOpenHandles` to troubleshoot this issue.
> ```
> 
> Keep in mind I commented out the `startStreamingServer()` function in `app.js`, if that's the reason why, then ignore this.
> 
> ### **Summary**
> This review does not really cover the streaming or messaging functions of this PR. However, I think the authentication looks good enough for now. Some things to consider:
> 
> 1. Make sure to add .env to gitignore
> 2. Maybe consider removing some console.logs
> 3. Consider which routes are going to need authentication and which are not
> 
> Keep in mind this is not fully my area of expertise, but hopefully something here was helpful!

Thank you for your review! What I have done:
1. Fixed the test issue by closing the server after all tests are done
2. Added a ticket for console.log 
3. Fixed the image router to use auth and also the base route
4. I could not replicate the error on images when sending token from login
5. Deleted .env from github
